### ISSUE TITLE
Anomaly "Uncaught exception Retyping.RetypeError(0)" during extraction (a Mathcomp project)

### ISSUE THREAD

**Issue by anton-trunov** [coq/coq#13581]
#### Description of the problem

The following snippet produces an anomaly `"Uncaught exception Retyping.RetypeError(0)."` when trying to extract it. This came up in an attempt to extract a Mathcomp-based project. The original code was minimized by @dmitromikh using [coq-bug-minimizer](https://github.com/JasonGross/coq-tools#coq-bug-minimizer) and then I minimized it even more by hand :). I can't seem to remove any more code from the snippet: the parameter `b` and `T := T0` seem to be essential here.

```coq
From Coq Require Extraction.

Record mixin_of T0 (b : unit) (T := T0) := Mixin { _ : T0 }.
Fail Extraction mixin_of.
```

Here is a stack trace generated with `coqc -q -debug anomaly_extraction.v`:

```
File "./anomaly_extraction.v", line 4, characters 0-20:
Error: Anomaly "Uncaught exception Retyping.RetypeError(0)."
Please report at http://coq.inria.fr/bugs/.
frame @ file "toplevel/coqc.ml", line 44, characters 2-81
frame @ file "toplevel/ccompile.ml", line 211, characters 12-37
frame @ file "toplevel/ccompile.ml", line 144, characters 18-89
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "toplevel/vernac.ml", line 95, characters 15-501
frame @ file "lib/flags.ml", line 24, characters 4-26
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "toplevel/vernac.ml", line 110, characters 23-65
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "toplevel/vernac.ml", line 68, characters 31-52
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "stm/stm.ml", line 2687, characters 4-60
frame @ file "stm/stm.ml", line 2561, characters 4-105
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "stm/stm.ml", line 1006, characters 6-10
frame @ file "stm/stm.ml", line 1086, characters 4-58
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "vernac/vernacinterp.ml", line 262, characters 18-43
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "lib/flags.ml", line 17, characters 14-17
frame @ file "vernac/vernacinterp.ml", line 218, characters 20-60
frame @ file "vernac/vernacinterp.ml", line 28, characters 19-23
frame @ file "plugins/extraction/extract_env.ml", line 654, characters 43-68
frame @ file "plugins/extraction/extract_env.ml", line 427, characters 10-79
frame @ file "plugins/extraction/extract_env.ml", line 333, characters 21-47
frame @ file "plugins/extraction/extraction.ml", line 1209, characters 12-30
raise @ unknown
frame @ file "plugins/extraction/extraction.ml", line 404, characters 7-36
frame @ file "plugins/extraction/extraction.ml", line 476, characters 12-77
frame @ file "plugins/extraction/extraction.ml", line 562, characters 8-39
frame @ file "plugins/extraction/extraction.ml", line 49, characters 2-70
frame @ file "pretyping/retyping.ml", line 206, characters 19-58
raise @ file "pretyping/retyping.ml", line 51, characters 22-44
```

#### Coq Version

about two week old dev (5fce39e0078935662fd6fb9b8e2ee6b0da60b3ab), 8.12.1, 8.11.1, 8.10.2, 8.9.1

**Comment by herbelin** [comment]
@anton-trunov: Thanks a lot for reporting!

@pi8027: just to be sure that we are not two to do the same job, would you like to look at this issue or are you currently busy with other stuff, in which case I'm ok to have a look? (It seems like a confusion between the length of the context of parameters and the expected number of parameters.)

**Comment by pi8027** [comment]
@herbelin I don't intend to investigate this issue very soon. So you may take a look. Thanks!

**Comment by herbelin** [comment]
Thanks for the constructive interaction! I'll try to propose something.

**Comment by herbelin** [comment]
Fixed in #13624.
